### PR TITLE
Add chunk exclusion for space dimensions to UPDATE/DELETE

### PR DESCRIPTION
**Pull request by svenklemm**
This patch adds support for chunk exclusion to UPDATE/DELETE on
space dimensions of hypertables.

### CODE REVIEW

**Comment by svenklemm** [comment]
#4374 should be merged before this

**Comment by fabriziomello** [comment]
Seems you added the generated test sql files insted of the related output files for each supported postgres version

File: src/planner/space_constraint.c [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+	partcall->args = list_make1(copyObject(var));
+
+	OpExpr *ret = (OpExpr *) make_opclause(tce->eq_opr /* opno */,
+										   BOOLOID /*opresulttype */,
+										   false /* opretset */,
+										   (Expr *) partcall /* left */,
+										   (Expr *) part_value /* right */,
+										   InvalidOid /* opcollid */,
+										   InvalidOid /* inputcollid */);
+	ret->location = -29811;
```
Comment (akuzm) [inline src/planner/space_constraint.c:150]: Should be a macro I guess

File: src/planner/space_constraint.c [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+	op2->args = list_make2(partcall, arr2);
+	op2->inputcollid = InvalidOid;
+	op2->useOr = true;
+	op2->location = PLANNER_LOCATION_MAGIC;
+
+	return op2;
+}
+
+Node *
+ts_amend_space_constraints(PlannerInfo *root, List *rtable, Node *node)
```
Comment (akuzm) [inline src/planner/space_constraint.c:205]: Let's add a top-level description of the algorithm here. If we have an equality condition on a space partitioning column, we also add a corresponding condition on `get_partition_hash` on this column. These conditions match the constraints on chunks, so Postgres' constraint exclusion is able to use them and exclude the chunks. Did I get it right?

`amend` sounds slightly imprecise, makes me think that we modify something, but here we just add a different form that matches the constraints on chunks. Maybe `ts_add_space_constraints`? Anyway, not so important.

File: tsl/test/expected/dist_hypertable-12.out [hunk 3]
```diff
@@ -239,91 +239,43 @@ UPDATE disttable SET temp = 3.7 WHERE device = 1;
...
    Foreign Update on _timescaledb_internal._dist_hyper_1_4_chunk
      Remote SQL: UPDATE _timescaledb_internal._dist_hyper_1_4_chunk SET temp = $2 WHERE ctid = $1
-   Foreign Update on _timescaledb_internal._dist_hyper_1_5_chunk
-     Remote SQL: UPDATE _timescaledb_internal._dist_hyper_1_5_chunk SET temp = $2 WHERE ctid = $1
-   Foreign Update on _timescaledb_internal._dist_hyper_1_6_chunk
-     Remote SQL: UPDATE _timescaledb_internal._dist_hyper_1_6_chunk SET temp = $2 WHERE ctid = $1
    ->  Seq Scan on public.disttable
          Output: disttable."time", disttable.device, NULL::integer, '3.7'::double precision, disttable.ctid
-         Filter: (disttable.device = 1)
+         Filter: ((disttable.device = 1) AND (_timescaledb_internal.get_partition_hash(disttable.device) = 242423622))
```
Comment (akuzm) [inline tsl/test/expected/dist_hypertable-12.out:246]: Do you want to delete them back as well as you did with constraints for `now()`?

**Review by akuzm — APPROVED** [review]
lgtm

File: src/planner/space_constraint.c (hunk 2 above)
Comment (akuzm) [reply src/planner/space_constraint.c:205]: TBH I would expect the Postgres planner to be smart enough to prove that, if you have an immutable function in a constraint and a constant condition in WHERE...

File: tsl/test/expected/dist_hypertable-12.out (hunk 3 above)
Comment (svenklemm) [reply tsl/test/expected/dist_hypertable-12.out:246]: yes they shouldnt end up in the explain output, will fix this

File: src/planner/space_constraint.c (hunk 2 above)
Comment (akuzm) [reply src/planner/space_constraint.c:205]: `operator_predicate_proof` doesn't evaluate sides of the constraint and only matches costraint & clause verbatim. And `eval_const_expression` we do on the relation constraints can't use equivalence classes to evaluate Vars to const. So no, it's not smart enough :(

**Review by gayyappan — APPROVED** [review]
Minor comments.
Changes look good to me.

File: src/planner/space_constraint.c [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+	arr2->multidims = false;
+	arr2->location = -1;
+	arr2->elements = part_values;
+
+	ScalarArrayOpExpr *op2 = makeNode(ScalarArrayOpExpr);
+	op2->opno = tce->eq_opr;
+	op2->args = list_make2(partcall, arr2);
+	op2->inputcollid = InvalidOid;
+	op2->useOr = true;
+	op2->location = PLANNER_LOCATION_MAGIC;
```
[the remark this answers is not in the document]
Comment (svenklemm) [reply src/planner/space_constraint.c:199]: This is so we can easily spot which nodes we added here so we can easily remove them once they are no longer needed.

[the remark this answers is not in the document]
Comment (svenklemm) [reply src/planner/space_constraint.c:199]: relevant code is in src/planner/constraint_cleanup.c

File: src/planner/space_constraint.c (hunk 2 above)
Comment (svenklemm) [reply src/planner/space_constraint.c:205]: done

File: tsl/test/expected/dist_hypertable-12.out (hunk 3 above)
Comment (svenklemm) [reply tsl/test/expected/dist_hypertable-12.out:246]: we could make the constraint removal a guc but i'm somewhat hesitant to add a GUC like that

File: src/planner/space_constraint.c [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+	{
+		Dimension *dim = &ht->space->dimensions[i];
+		if (dim->type == DIMENSION_TYPE_CLOSED && dim->column_attno == varattno)
+		{
+			return dim;
+		}
+	}
+	return NULL;
+}
+
```
[the remark this answers is not in the document]
Comment (svenklemm) [reply src/planner/space_constraint.c:44]: done

File: src/planner/space_constraint.c [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+{
+	/* build FuncExpr to use in eval_const_expressions */
+	return makeFuncExpr(partfunc.func_fmgr.fn_oid /* funcid */,
+						partfunc.rettype /* rettype */,
+						args /* args */,
+						InvalidOid /* funccollid */,
+						inputcollid /* inputcollid */,
+						COERCE_EXPLICIT_CALL /* fformat */);
+}
+
```
[the remark this answers is not in the document]
Comment (svenklemm) [reply src/planner/space_constraint.c:120]: done

**Pull request merged**